src/pullpush.c

```diff
@@ -0,0 +1,153 @@
+#include "internal.h"
+#include "script_int.h"
+
+#include <errno.h>
+#include <limits.h>
+#include <stdbool.h>
+#include "pullpush.h"
+
+unsigned char *push_bytes(unsigned char **cursor, size_t *max,
+                          const void *src, size_t len)
+{
+    if (cursor == NULL || *cursor == NULL) {
+        *max += len;
+        return NULL;
+    }
+    if (len > *max) {
+        if (src)
+            memcpy(*cursor, src, *max);
+        /* From now on, max records the room we *needed* */
+        *max = len - *max;
+        *cursor = NULL;
```

Comment:
What's the purpose of doing this on failure?

Reply:
*cursor == NULL is the way you detect unmarshalling failure.

*max increasing is how you tell how many bytes short you are.

You'll see this explicitly used here:

```
    if (cursor == NULL) {
            /* Once cursor was NULL, max accumulates hm bytes we needed */
            *bytes_written = len + max;
            return WALLY_EINVAL;
    } else {
            *bytes_written = len - max;
    }
```